File: 26-divide-backtrack/1-intro/main.cpp
// Divide and Conquer & Backtracking - C++ example

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// Merge function for merge sort
vector<int> merge(const vector<int>& left, const vector<int>& right) {
    vector<int> result;
    int i = 0, j = 0;
    while (i < (int)left.size() && j < (int)right.size()) {
        if (left[i] < right[j]) result.push_back(left[i++]);
        else result.push_back(right[j++]);
    }
    while (i < (int)left.size()) result.push_back(left[i++]);
    while (j < (int)right.size()) result.push_back(right[j++]);
    return result;
}

// Merge sort using divide and conquer
vector<int> mergeSort(const vector<int>& arr) {
    if (arr.size() <= 1) return arr;
    int mid = arr.size() / 2;
    vector<int> left(arr.begin(), arr.begin() + mid);
    vector<int> right(arr.begin() + mid, arr.end());
    return merge(mergeSort(left), mergeSort(right));
}

// Backtracking subset sum
bool subsetSum(const vector<int>& nums, int target, int index = 0) {
    if (target == 0) return true;
    if (target < 0 || index >= (int)nums.size()) return false;
    // Include nums[index]
    if (subsetSum(nums, target - nums[index], index + 1)) return true;
    // Exclude nums[index]
    return subsetSum(nums, target, index + 1);
}

int main() {
    vector<int> arr = {3, 6, 1, 7, 2, 5};
    vector<int> sortedArr = mergeSort(arr);
    cout << "Sorted array: ";
    for (int x : sortedArr) cout << x << ' ';
    cout << endl;

    vector<int> nums = {3, 34, 4, 12, 5, 2};
    int target = 9;
    cout << "Subset sum exists: " << boolalpha << subsetSum(nums, target) << endl;

    return 0;
}

File: 26-divide-backtrack/1-intro/main.java
// Divide and Conquer & Backtracking - Java example

import java.util.*;

class Main {
    // Merge function for merge sort
    public static List<Integer> merge(List<Integer> left, List<Integer> right) {
        List<Integer> result = new ArrayList<>();
        int i = 0, j = 0;
        while (i < left.size() && j < right.size()) {
            if (left.get(i) < right.get(j)) result.add(left.get(i++));
            else result.add(right.get(j++));
        }
        while (i < left.size()) result.add(left.get(i++));
        while (j < right.size()) result.add(right.get(j++));
        return result;
    }

    // Merge sort using divide and conquer
    public static List<Integer> mergeSort(List<Integer> arr) {
        if (arr.size() <= 1) return arr;
        int mid = arr.size() / 2;
        return merge(mergeSort(arr.subList(0, mid)), mergeSort(arr.subList(mid, arr.size())));
    }

    // Backtracking subset sum
    public static boolean subsetSum(List<Integer> nums, int target, int index) {
        if (target == 0) return true;
        if (target < 0 || index == nums.size()) return false;

        // Include nums.get(index)
        if (subsetSum(nums, target - nums.get(index), index + 1)) return true;

        // Exclude nums.get(index)
        return subsetSum(nums, target, index + 1);
    }

    public static void main(String[] args) {
        List<Integer> arr = Arrays.asList(3, 6, 1, 7, 2, 5);
        List<Integer> sortedArr = mergeSort(arr);
        System.out.println("Sorted array: " + sortedArr);

        List<Integer> nums = Arrays.asList(3, 34, 4, 12, 5, 2);
        int target = 9;
        System.out.println("Subset sum exists: " + subsetSum(nums, target, 0));
    }
}
